opensquirrel/circuit_builder.py

```diff
@@ -43,27 +42,21 @@ class CircuitBuilder(InstructionLibrary):
 
     """
 
-    def __init__(
-        self,
-        qubit_register_size: int,
-        bit_register_size: int = 0,
-        gate_set: Mapping[str, Callable[..., Gate]] = default_gate_set,
-        non_unitary_set: Mapping[str, Callable[..., NonUnitary]] = default_non_unitary_set,
-    ) -> None:
-        InstructionLibrary.__init__(self, gate_set, non_unitary_set)
+    def __init__(self, qubit_register_size: int, bit_register_size: int = 0) -> None:
         self.register_manager = RegisterManager(QubitRegister(qubit_register_size), BitRegister(bit_register_size))
         self.ir = IR()
 
     def __getattr__(self, attr: Any) -> Callable[..., Self]:
         return partial(self._add_instruction, attr)
 
     def _add_instruction(self, attr: str, *args: Any) -> Self:
-        if attr in self.gate_set:
-            generator_f_gate = self.get_gate_f(attr)
+        instruction_library = InstructionLibrary()
```

Comment:
That's right. `InstructionLibrary()` ends up executing `Singleton.__call__` which is where you say:

- Do I have an instance of InstructionLibrary already? Otherwise I create that single instance first.
- Then I return the single instance.

Reply:
In C++ I always implement Singleton's using a `get_instance` static method. So you would do something like `InstructionLibrary.get_instance()` instead of `InstructionLibrary()`. If you also think that is clearer, we can change the code to do that. `get_instance` implementation would be the same than that of `__call__`.

Reply:
A `module` can generally be considered a singleton in Python, so we could also have the instruction library as a module and not implement it as a class.

E.g. `instruction_library.py`:
```
from __future__ import annotations

from collections.abc import Callable, Mapping
from typing import TYPE_CHECKING

from opensquirrel.default_instructions import default_gate_set, default_non_unitary_set

if TYPE_CHECKING:
    from opensquirrel.ir import Gate, NonUnitary


gate_set = default_gate_set
non_unitary_set = default_non_unitary_set

def get_gate_f(gate_name: str) -> Callable[..., Gate]:
    if gate_name not in gate_set:
        msg = f"unknown gate '{gate_name}'"
        raise ValueError(msg)
    return gate_set[gate_name]

def get_non_unitary_f(non_unitary_name: str) -> Callable[..., NonUnitary]:
    if non_unitary_name not in non_unitary_set:
        msg = f"unknown non-unitary instruction '{non_unitary_name}'"
        raise ValueError(msg)
    return non_unitary_set[non_unitary_name]
```
This can then be imported into other modules:

```
import instruction_library

...
gate_set = instruction_library.gate_set
...
generator_f_gate = instruction_library.get_gate_f(attr)
...

Reply:
As a matter of fact, there are many ways to implement this controversial design pattern in Python. I just read in a couple of places that the metaclass approach was the preferred way (one of those places, [here](https://stackoverflow.com/questions/6760685/what-is-the-best-way-of-implementing-singleton-in-python), where the second most voted answer is to use a module).

I personally see these pros and cons to each solution.

### Metaclass
- Pros:
  - I like the more object-oriented solution of using an InstructionLibrary class (encapsulation, consistency with the rest of the code, where we have Circuit's and CircuitBuilder's, and RegisterManager's, all classes).
  - I like the explicitness of declaring `class InstructionLibrary(metaclass=Singleton)`.
- Cons:
  - `Singleton` class implementation is complex, mainly due to type annotation.
  - I would very much prefer to be able to get an instance of a singleton class via `get_instance()` than via `__call__`.

### Module
- Pros:
  - Really simple. You get rid of the `utils/singleton.py`, the `Singleton` class... everything.
- Cons:
  - I find it confusing to read `instruction_library.gate_set` and think `instruction_library` is an object, while it really is a module.

Reply:
Done. You beat me. I cannot say no to removing code, especially complex code. 😀

Reply:
Thanks for looking into this in detail! Great to know about the alternatives 😃